plugins.md

```diff
@@ -0,0 +1,225 @@
+module to generate trace spans when that module is used. We provide plugins for
+some well-known modules such as express, mongodb, and http, and provide a plugin
+API for developers to create their own.
+
+## Outline
+
+A plugin consists of a set of *patch objects*. A patch object gives information
+about how a file in a module should be patched in order to create trace spans.
+
+Each patch object should contain the following fields:
+
+* `file`: The path to the file whose exports should be patched, relative to the
+  root of the module. You can specify an empty string, or omit this field
+  entirely, to specify the export of the module itself.
+* `versions`: A `semver` expression which will be used to control whether the
+  specified file will be patched based on the module version; the patch will
+  only be applied if the loaded module version satisfies this expression. This
+  might be useful if your plugin only works on some versions of a module, or if
+  you are patching internal mechanisms that are specific to a certain range of
+  versions.
+* `patch`: A function describing how the given file should be patched. It will
+  be passed two arguments: the object exported from the file, and an object that
+  exposes functions for creating spans and propagating context (see
+  [What the API provides](#what-the-api-provides) for more details).
+
+A plugin simply exports a list of patch objects.
+
+For example, here's what a plugin for `express` might export:
+
+```js
+// express 4.x
+
+function patchModuleRoot(expressModule, api) {
+  // ...
+}
+
+module.exports = [
+  { file: '', versions: '4.x', patch: patchModuleRoot }
+];
+```
+
+In most cases, it should be sufficient to patch just the root (public exports)
+of the module, in which case the `file` field can set to `''` or omitted. Based
+on how the module being patched is implemented, however, it may be necessary to
+patch other parts of the module as well.
+
+### Enabling plugins in the Trace Agent
+
+Developers wishing to trace their applications may specify, in the
+[configuration][config-js] object passed to the trace agent, a `plugins` field
+with a key-value pair (module name, path to plugin). For
+example, to patch just `express` with a plugin at `./plugins/plugin-express.js`:
+
+require('@google-cloud/trace-agent').start({
+  plugins: {
+    express: path.join(__dirname, 'plugins/express.js')
+  }
+})
+
+At this time, the path to the plugin may either be a module name or an absolute
+path to the plugin.
+
+## What the API provides
+
+The `api` object, in short, provides functions that facilitate the following:
+
+- Creating trace spans and add labels to them.
+- Getting information about how the trace agent was configured in the
+  current application.
+- Parsing and serializing trace contexts for the sake of propagating them over
+  the network.
+- Binding callbacks and event emitters with a CLS namespace in order to
```

Comment:
You don't need to mention the mechanism (CLS) since we may want to change that in the future. 'Binding callbacks and event emitters in order to propagate...'